AX: move AXObjectCache::postNotification code inside the accessibility directory

minorninth opened the pull request:
#### 89801b35ee043e7e168adc2c19bb5d14704a7910
<pre>
AX: move AXObjectCache::postNotification code inside the accessibility directory
<a href="https://bugs.webkit.org/show_bug.cgi?id=297150">https://bugs.webkit.org/show_bug.cgi?id=297150</a>
<a href="https://rdar.apple.com/157888495">rdar://157888495</a>

Reviewed by Tyler Wilcock.

Knowing which accessibility notification to fire is logic that belongs
in the accessibility module. Let&apos;s take the small amount of code that
posts a notification from somewhere else in WebCore, and change it to
call a specific, narrow function on AXObjectCache, and let
AXObjectCache post the notification.

As a follow-up, we&apos;ll be able to reduce the number of accessibility
files that much of the codebase needs to import.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onDraggingStarted):
(WebCore::AXObjectCache::onDraggingEnded):
(WebCore::AXObjectCache::onDraggingEnteredDropZone):
(WebCore::AXObjectCache::onDraggingExitedDropZone):
(WebCore::AXObjectCache::onDraggingDropped):
(WebCore::AXObjectCache::onTopDocumentLoaded):
(WebCore::AXObjectCache::onNonTopDocumentLoaded):
(WebCore::AXObjectCache::onAutocorrectionOccured):
(WebCore::AXObjectCache::onEditableTextValueChanged):
(WebCore::AXObjectCache::onDocumentInitialFocus):
(WebCore::AXObjectCache::onLayoutComplete):
(WebCore::AXObjectCache::onSelectedOptionChanged):
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::onSelectedChanged): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::implicitClose):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::respondToChangedContents):
(WebCore::Editor::markAndReplaceFor):
* Source/WebCore/html/HTMLOptionElement.cpp:
(WebCore::HTMLOptionElement::setSelectedState):
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::setInnerTextValue):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::dispatchDragEvent):
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::setInitialFocus):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::didLayout):
* Source/WebCore/rendering/RenderMenuList.cpp:
(RenderMenuList::didUpdateActiveOption):

Canonical link: <a href="https://commits.webkit.org/298476@main">https://commits.webkit.org/298476@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b8c3d7e74ad79df2dc7abd3683d500fef5fdbc2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115598 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35261 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25784 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121645 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66110 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117487 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35942 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43844 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87810 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118546 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28668 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103743 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68208 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27804 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21859 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65324 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98043 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21980 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124794 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42493 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31853 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96567 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42860 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99933 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96353 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24525 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41606 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19462 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38363 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42384 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47956 "Built successfully") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41858 "Built successfully") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45188 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43571 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -1301,8 +1301,8 @@ void LocalFrameView::didLayout(SingleThreadWeakPtr<RenderElement> layoutRoot, bo
 #if PLATFORM(COCOA) || PLATFORM(WIN) || PLATFORM(GTK)
-    if (CheckedPtr cache = document->existingAXObjectCache())
-        cache->postNotification(layoutRoot.get(), AXNotification::LayoutComplete);
+    if (auto* cache = document->existingAXObjectCache())
```
This should still be CheckedPtr, will fix momentarily...

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -2065,6 +2122,13 @@ void AXObjectCache::onSelectedChanged(Element& element)
+void AXObjectCache::onSelectedOptionChanged(RenderObject& menuList, int optionIndex)
+{
+    auto* axObject = get(menuList);
+    if (RefPtr menuList = dynamicDowncast<AccessibilityMenuList>(axObject))
```
The Safer-CPP static analyzer might complain about this raw pointer (but maybe not?)

Do you feel the local variable helps readability enough to be worth it? IMO this isn't so bad:

if (RefPtr menuList = dynamicDowncast<AccessibilityMenuList>(get(menuList)))

twilco commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -576,8 +576,8 @@ bool FocusController::setInitialFocus(FocusDirection direction, KeyboardEvent* p
     RefPtr focusedOrMainFrame = this->focusedOrMainFrame();
-    if (CheckedPtr cache = focusedOrMainFrame ? focusedOrMainFrame->document()->existingAXObjectCache() : nullptr)
-        cache->postNotification(focusedOrMainFrame->document(), AXNotification::FocusedUIElementChanged);
+    if (auto* cache = focusedOrMainFrame ? focusedOrMainFrame->document()->existingAXObjectCache() : nullptr)
```
Should this one also be CheckedPtr?

twilco reviewed (approved):
Love it 🤩

minorninth committed:
AX: move AXObjectCache::postNotification code inside the accessibility directory
https://bugs.webkit.org/show_bug.cgi?id=297150
rdar://157888495

Reviewed by Tyler Wilcock.

Knowing which accessibility notification to fire is logic that belongs
in the accessibility module. Let's take the small amount of code that
posts a notification from somewhere else in WebCore, and change it to
call a specific, narrow function on AXObjectCache, and let
AXObjectCache post the notification.

As a follow-up, we'll be able to reduce the number of accessibility
files that much of the codebase needs to import.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onDraggingStarted):
(WebCore::AXObjectCache::onDraggingEnded):
(WebCore::AXObjectCache::onDraggingEnteredDropZone):
(WebCore::AXObjectCache::onDraggingExitedDropZone):
(WebCore::AXObjectCache::onDraggingDropped):
(WebCore::AXObjectCache::onTopDocumentLoaded):
(WebCore::AXObjectCache::onNonTopDocumentLoaded):
(WebCore::AXObjectCache::onAutocorrectionOccured):
(WebCore::AXObjectCache::onEditableTextValueChanged):
(WebCore::AXObjectCache::onDocumentInitialFocus):
(WebCore::AXObjectCache::onLayoutComplete):
(WebCore::AXObjectCache::onSelectedOptionChanged):
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::onSelectedChanged): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::implicitClose):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::respondToChangedContents):
(WebCore::Editor::markAndReplaceFor):
* Source/WebCore/html/HTMLOptionElement.cpp:
(WebCore::HTMLOptionElement::setSelectedState):
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::setInnerTextValue):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::dispatchDragEvent):
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::setInitialFocus):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::didLayout):
* Source/WebCore/rendering/RenderMenuList.cpp:
(RenderMenuList::didUpdateActiveOption):

Canonical link: https://commits.webkit.org/298476@main
